### ISSUE TITLE
Strange behavior when using limit() with certain values.

### ISSUE THREAD

**Issue by thohag** [Automattic/mongoose#4202]
Using mongoose 4.4.19

Not sure what is actually causing this, but I supply code to recreate the problem I got here:

I want mongodb to reconnect when connection is dropped, but the limit function seems to change the reconnection behavior depending on the value supplied.

The following case works as expected:

```
var mongoose = require('mongoose');
var options = { server: { socketOptions: { keepAlive: 1, autoReconnect: true }, reconnectTries: 600, reconnectInterval: 1000 } };
mongoose.connect(uri, options);
var testsShema = new mongoose.Schema({},{ strict: false });
var tests = mongoose.model('tests', testsShema);
var inserted = false;
mongoose.connection.on("open", function(){
    console.log("Connected")
    if (inserted) {return;}
    return insertDocuments();
});
mongoose.connection.on("disconnected", function(){ console.log("Disconnected")});
mongoose.connection.on("reconnected", function(){console.log("Reconnected")});
function insertDocuments() {
    inserted = true;
    var bulk = tests.collection.initializeUnorderedBulkOp();
    for (var i=0;i<10;i++) {bulk.insert({test:i});}
    bulk.execute({w:1}, function(resp) {queryDocuments();});
}
function queryDocuments() {
    var limit = 1;
    var query = tests.find({}).limit(limit);
    query.exec(function(err, docs) {
       console.log("Got",docs.length); 
    });
}
process.on('exit', function(){console.log("Process closing")});
```
After running this code i close mongod and then start it again,
the output from the script above is what I expect:

```
Connected
Got 1
Disconnected
Reconnected
```

I then change limit variable in the queryDocuments function to 5:

```
function queryDocuments() {
    var limit = 5;
    var query = tests.find({}).limit(limit);
    query.exec(function(err, docs) {
       console.log("Got",docs.length); 
    });
}
```

I then run the script again and restart mongod, but the result is that node exits when mongodb is closed:

```
Connected
Got 5
Process closing
```

Is this expected behavior?

If I change the limit value to something like 100, I get the expected behavior again:

```
Connected
Got 10
Disconnected
Reconnected
```

Again, this is what I expect, thus it seems that the limit operator causes different outcomes depending on the value supplied.

(between tests I remove all documents in the collection)

Setting the limit value to -5 seems to alleviate the problem (behavior is as expected), but it does not seem like an optimal solution as it may return less documents than requested, from the mongodb docs: 

> A negative limit is similar to a positive limit but closes the cursor after returning a single batch of results. As such, with a negative limit, if the limited result set does not fit into a single batch, the number of documents received will be less than the specified limit. By passing a negative limit, the client indicates to the server that it will not ask for a subsequent batch via getMore.

**Comment by vkarpov15** [comment]
Hmm which version of node and which version of mongodb are you using? Below script (same as yours with a couple differences at the beginning) works fine for me:

```javascript
'use strict';

var assert = require('assert');
var mongoose = require('mongoose');
var Schema = mongoose.Schema;

mongoose.set('debug', true);

var options = { server: { socketOptions: { keepAlive: 1, autoReconnect: true }, reconnectTries: 600, reconnectInterval: 1000 } };
mongoose.connect('mongodb://localhost/gh4202', options);

var testsShema = new mongoose.Schema({},{ strict: false });
var tests = mongoose.model('tests', testsShema);
var inserted = false;
mongoose.connection.on("open", function(){
    console.log("Connected")
    if (inserted) {return;}
    return insertDocuments();
});
mongoose.connection.on("disconnected", function(){ console.log("Disconnected")});
mongoose.connection.on("reconnected", function(){console.log("Reconnected")});
function insertDocuments() {
    inserted = true;
    var bulk = tests.collection.initializeUnorderedBulkOp();
    for (var i=0;i<10;i++) {bulk.insert({test:i});}
    bulk.execute({w:1}, function(resp) {queryDocuments();});
}
function queryDocuments() {
    var limit = 5;
    var query = tests.find({}).limit(limit);
    query.exec(function(err, docs) {
       console.log("Got",docs.length); 
    });
}
process.on('exit', function(){console.log("Process closing")});
```

Output:

```
$ node gh-4202.js 
Connected
Mongoose: tests.initializeUnorderedBulkOp()   
Mongoose: tests.find({}) { limit: 1, fields: undefined }  
Got 1
Disconnected
Reconnected
^C
$ 
$ 
$ 
$ node gh-4202.js 
Connected
Mongoose: tests.initializeUnorderedBulkOp()   
Mongoose: tests.find({}) { limit: 5, fields: undefined }  
Got 5
Disconnected
Reconnected
^C
```

In both cases, I run the script and then kill the mongodb process and restart it. When are you stopping mongodb? Are you stopping it during the query?

**Comment by thohag** [comment]
My apologies for not reporting more on the environment in the first post,
db version v3.0.3 (mongod --version)
node versions tested: v0.12.1 v4.4.5 v6.2.0
Windows 10 (x64)

Test with your code on my system:

>node test2
Connected
Mongoose: tests.initializeUnorderedBulkOp()
Mongoose: tests.find({}) { limit: 1, fields: undefined }
Got 1
Disconnected
Reconnected


>node test2
Connected
Mongoose: tests.initializeUnorderedBulkOp()
Mongoose: tests.find({}) { limit: 5, fields: undefined }
Got 5
Process closing

I start the script, wait for the "Got 5" message, then "ctrl+c" the mongod process and restart it by running mongod again. I am not stopping it during the query.

**Comment by thohag** [comment]
I did some further tests with different mongodb versions on windows (no ssl):

|Version|Disconnect event	|Reconnect event	|Process exited |
|------------------|---------------------|-------------------------------|-----------------------------|
|3.3.8|FALSE	|TRUE	|FALSE|
|3.2.7|TRUE	|TRUE	|FALSE|
|3.1.9|TRUE	|TRUE	|FALSE|
|3.1.8|TRUE	|TRUE	|FALSE|
|3.1.7|TRUE	|TRUE	|FALSE|
|3.1.6|TRUE	|TRUE	|FALSE|
|3.1.5|TRUE	|TRUE	|FALSE|
|3.1.4|FALSE	|FALSE	|TRUE|
|3.1.0|FALSE	|FALSE	|TRUE|
|3.0.3|FALSE	|FALSE	|TRUE|

The problem seems to go away with version 3.1.5, but something seems to happen with the disconnect event at version 3.3.8 (git version: f2f012ccc781b1d3978cb5e425d6747a0b204b04).

**Comment by vkarpov15** [comment]
So mongodb versions 3.1.x and 3.3.x are unstable development versions, I don't think the underlying mongodb driver has official support for them, so those aren't under consideration. Looks like mongodb 3.2 works as expected for you but 3.0 does not?

**Comment by thohag** [comment]
Yes, 3.2 works but 3.0 does not. Are you able to reproduce with 3.0.3 or similar?

**Comment by vkarpov15** [comment]
So looks like this was fixed by https://github.com/Automattic/mongoose/commit/3d161ecabf6a8706518696225712679efefd8cba, with mongodb-core 1.3.21: https://github.com/christkv/mongodb-core/blob/2.0/HISTORY.md#1321-2016-05-30 . Will be fixed once I release 4.5.4 today :+1:
